src/main/java/org/jenkinsci/plugins/workflow/support/steps/ExecutorStepExecution.java

```diff
@@ -480,6 +480,71 @@ public String getCookie() {
             return getDisplayName();
         }
 
+        static String findLabelName(FlowNode flowNode){
+            LabelAction la = flowNode.getPersistentAction(LabelAction.class);
+
+            if (la != null) {
+                return la.getDisplayName();
+            }
+            return null;
+        }
+
+        /**
+         * Similar to {@link #getEnclosingLabel()}.
+         * However instead of returning the innermost label including labels inside node blocks this one
+         * concatenates all labels found outside the current (node) block
+         *
+         * As {@link FlowNode#getEnclosingBlocks()} will return the blocks sorted from inner to outer blocks
+         * this method will create a string like
+         * <code>#innerblock#outerblock for</code> for a script like
+         * <pre>
+         *     {@code
+         *     parallel(outerblock: {
+         *         stage('innerblock') {
+         *             node {
+         *                 // .. do something here
+         *             }
+         *         }
+         *     }
+         *     }
+         * </pre>
+         *
+         * In case there's no context available or we get a timeout we'll just return <code>baseLabel</code>
+         *
+         * */
+        private String concatenateAllEnclosingLabels(StringBuilder labelName) {
+            if (!context.isReady()) {
+                return labelName.toString();
+            }
+            FlowNode executorStepNode = null;
+            try (Timeout t = Timeout.limit(100, TimeUnit.MILLISECONDS)) {
```

Comment:
When or how often is it expected to run into this timeout?
I am just wondering if this will result in many thrown exceptions on a system under load put it under even more stress.

Reply:
The timeout was taken from the code used to find the enclosing label for displaying in the UI.
I do not expect it to happen frequently. We have a big system (node allocation itself sometimes takes more than a minute - even for Freestyle jobs and pipeline jobs BEFORE applying this patch) and never got any timeout here.
Hope that helps.
Since this was done analogous to what is done for the UI display maybe one of the maintainers can comment on this, too?

Reply:
AFAIU the timeout was added to the UI code in [this commit](https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/2/commits/771d3aa57e65330b703e82abe2f9835471676f71) because `context.get(FlowNode.class)` could block before [this change](https://github.com/jenkinsci/workflow-cps-plugin/pull/145) in workflow-cps. Based on a quick read of the discussion in #2 I think the call is expected to be instantaneous, see [this comment](https://github.com/jenkinsci/workflow-durable-task-step-plugin/pull/2#discussion_r122758426). Maybe the timeout here is unnecessary.

Reply:
For UI code, certainly a timeout is in order—the exact result is not that important, we just need to complete the HTTP response, _now_.

For this code I am less sure, but I think it would be called in the context of `Queue.maintain`, and that also is expected to run in subsecond times. Whether a long delay is even theoretically possible here, I am not sure offhand; would have to take time to study it.